![](https://drive.google.com/uc?export=view&id=1L9JLLQHPZoMRwzYfmKcyM9VME_SHeZrr)

# TP 1 : pre-processing texts

In this practical session, we will see a few basic processing of textual data.

Within a computer, text is encoded as a string of characters.
In order to analyze textual data within NLP applications, we first need to properly preprocess it.
An NLP preprocessing pipeline generally consists of the following steps :
* sentence segmentation
* tokenisation
* normalization: lower-casing, lemmatization, optionally removing stop-words and punctuation
* pos-tagging
* named entity recognition
* parsing

The first two steps are necessary, while the others are optional.

For these exercises, we will mainly use the module **spacy** (already installed on google colab, but some additional libraries might be missing).

Spacy provides ways to carry out tasks such as segmentation, tokenization, lemmatization and pos-tagging.


We will also extract information from Wikipedia pages as an example.

## 0- Upload and read the text files

At first, we're going to use a text written in English. Then, we'll try to apply the tools to French.
We'll use the wikipedia library to extract pages from wikipedia

In [2]:
%pip install -q wikipedia


[notice] A new release of pip is available: 24.2 -> 24.3.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [3]:
import wikipedia
wikipedia.set_lang('en')
text_en = wikipedia.page("AdaLovelace")
print(text_en.content[:100])

Augusta Ada King, Countess of Lovelace (née Byron; 10 December 1815 – 27 November 1852), also known 


In [4]:
wikipedia.set_lang('fr')
text_fr = wikipedia.page("Lovelace")
print(text_fr.content[:100])

Ada Lovelace, de son nom complet Augusta Ada King, comtesse de Lovelace, née Ada Byron le 10 décembr


## 1- Using Spacy

All info about Spacy: https://spacy.io/ ; More info on the pipelines: https://spacy.io/usage/processing-pipelines

Spacy is a more realistic library for NLP than NLTK, with higher performances on the basic processing steps.

Spacy can be used to directly tokenize any text.
With spacy, we build a pipeline that does everything at once.
To make it work, you need to load a model specific to the target language, for example 'en' for English (there are also some domain specific models).


The model corresponds to a processing 'pipeline':
  by default, it includes the tokenisation, the lemmatization and the POS tagging

Using spacy:
- import the spacy module into Python
- load all the necessary models, e.g. for English



Then process a text with the pipeline:



In [47]:
import spacy
from spacy.tokens import Doc
import en_core_web_sm

nlp = spacy.load('en_core_web_sm')
doc = nlp(text_en.content)

# Access the vocabulary object
print("Vocabulary Object:")
print(doc.vocab)

lexeme = doc.vocab[doc.text]
print(f'lexeme. text  : {lexeme.text[0:65]},\nlexeme.orth : {lexeme.orth},\nlexeme.shape_ : {lexeme.shape_}')
# Returns the integer ID and returns the string if the ID exists
print(f'doc.vocab.strings["Byron"]: {doc.vocab.strings["Byron"]}\n') 
print(f'doc.vocab.strings[9562509122658583955] : {doc.vocab.strings[9562509122658583955]}')

Vocabulary Object:
lexeme. text  : Augusta Ada King, Countess of Lovelace (née Byron; 10 December 18,
lexeme.orth : 3869055741974767632,
lexeme.shape_ : LONG
doc.vocab.strings["Byron"]: 9562509122658583955

doc.vocab.strings[9562509122658583955] : Byron


In [6]:
# Iterate over the sentences
print("\nSentences:")
for sent in doc.sents:
    print(f'sent.text : {sent.text}')

# Iterate over the tokens
print("\nTokens:")
for token in doc:
    print(f"Token: {token.text}, POS: {token.pos_}, Lemma: {token.lemma_}, Entity Type: {token.ent_type_} dependency: {token.dep_}")

# Iterate over noun chunks
print("\nNoun Chunks:")
for chunk in doc.noun_chunks:
    print(f'chunk.text : {chunk.text}')

# Iterate over named entities
print("\nNamed Entities:")
for ent in doc.ents:
    print(f"Entity: {ent.text}, Label: {ent.label_}")


Sentences:
sent.text : Augusta Ada King, Countess of Lovelace (née Byron; 10 December 1815 – 27 November 1852), also known as Ada Lovelace, was an English mathematician and writer chiefly known for her work on Charles Babbage's proposed mechanical general-purpose computer, the Analytical Engine.
sent.text : She was the first to recognise that the machine had applications beyond pure calculation.

sent.text : Lovelace was the only legitimate child of poet Lord Byron and reformer Anne Isabella Milbanke.
sent.text : All her half-siblings, Lord Byron's other children, were born out of wedlock to other women.
sent.text : Lord Byron separated from his wife a month after Ada was born and left England forever.
sent.text : He died in Greece when she was eight.
sent.text : Lady Byron was anxious about her daughter's upbringing and promoted Lovelace's interest in mathematics and logic in an effort to prevent her from developing her father's perceived insanity.
sent.text : Despite this, Lovelace 

In [7]:
import spacy
from spacy.tokens import Doc
import fr_core_news_sm

nlp = spacy.load('fr_core_news_sm')
doc = nlp(text_fr.content)

# Access the vocabulary object
print("Vocabulary Object:")
print(doc.vocab)

lexeme = doc.vocab[doc.text]
print(f'lexeme. text  : {lexeme.text[0:65]},\nlexeme.orth : {lexeme.orth},\nlexeme.shape_ : {lexeme.shape_}')
# Returns the integer ID and returns the string if the ID exists
print(f'doc.vocab.strings["Augusta"]: {doc.vocab.strings["Byron"]}\n') 
print(f'doc.vocab.strings[9562509122658583955] : {doc.vocab.strings[9562509122658583955]}')

Vocabulary Object:
lexeme. text  : Ada Lovelace, de son nom complet Augusta Ada King, comtesse de Lo,
lexeme.orth : 18242613198961804001,
lexeme.shape_ : LONG
doc.vocab.strings["Augusta"]: 9562509122658583955

doc.vocab.strings[9562509122658583955] : Byron


Try to see what is stored with the document, especially doc.vocab, doc.sent and see what you get when you iterate over a document or its parts.

Hint: You can either access Spacy's manual on the internet to find out how to access the information, or look at the built-in help by typing help(doc). https://spacy.io/api/doc

 - The vocab object manages the mapping between strings and their corresponding integer IDs. 
 - You can access a Lexeme object using the string or its ID.
 - The vocab object also provides access to the StringStore, which maps strings to integers and vice versa

In [8]:
# Iterate over the sentences
print("\nSentences:")
for sent in doc.sents:
    print(f'sent.text : {sent.text}')

# Iterate over the tokens
print("\nTokens:")
for token in doc:
    print(f"Token: {token.text}, POS: {token.pos_}, Lemma: {token.lemma_}, Entity Type: {token.ent_type_}")

# Iterate over noun chunks
print("\nNoun Chunks:")
for chunk in doc.noun_chunks:
    print(f'chunk.text : {chunk.text}')

# Iterate over named entities
print("\nNamed Entities:")
for ent in doc.ents:
    print(f"Entity: {ent.text}, Label: {ent.label_}")


Sentences:
sent.text : Ada Lovelace, de son nom complet Augusta
sent.text : Ada King, comtesse de Lovelace, née Ada Byron le 10 décembre 1815 à Londres et morte le 27 novembre 1852 à Marylebone dans la même ville
sent.text : , est une pionnière de la science informatique.

sent.text : Elle est principalement connue pour avoir réalisé le premier véritable programme informatique, lors de son travail sur un ancêtre de l'ordinateur : la machine analytique de Charles Babbage.
sent.text : Dans ses notes, on trouve en effet le premier programme publié, destiné à être exécuté par une machine, ce qui fait d'Ada Lovelace la première personne au monde à avoir programmé.
sent.text : Elle a également entrevu et décrit certaines possibilités offertes par les calculateurs universels, allant bien au-delà du calcul numérique et de ce qu'imaginaient Babbage et ses contemporains.



sent.text : == Biographie ==



sent.text : ==
sent.text : =
sent.text : Environnement familial ===


sent.text : Ada est 

### 1.1 Tokenisation

**Exercise 1:** Tokenize the text in French
* Find a model for French and tokenize the text in the file. Hint: you will need to download the model first, that can be done in the notebook using: *spacy.cli.download( model_name )*
* What does contain the *doc* variable? Hint: You can either access Spacy's manual on the internet to find out how to access the information, or look at the built-in help by typing help(doc). https://spacy.io/api/doc
* Print the individual tokens. Do you see any error?
* How many tokens do we have?
* How many words / types / unique tokens do we have (i.e. vocabulary size)?
* Use Pandas to better visualize the results

In [ ]:
# Load the model
import spacy
import spacy.cli
spacy.cli.download("fr_core_news_sm")
nlp = spacy.load('fr_core_news_sm')

file_path = 'corpus_qa.txt'
with open( file_path  ) as infile:
  text = infile.read()

In [10]:
from spacy.tokens import Doc
import numpy as np
import fr_core_news_sm

doc = nlp(text)

# Access the vocabulary object
print("Vocabulary Object:")
print(doc.vocab)

# Preprocess using spacy's pipeline
# Iterate over the sentences
print("\nSentences:")
for sent in doc.sents:
    print(f'sent.text : {sent.text}')

# Iterate over the tokens
print("\nTokens:")
for token in doc:
    print(f"Token: {token.text}, POS: {token.pos_}, Lemma: {token.lemma_}, Entity Type: {token.ent_type_}")

# Iterate over noun chunks
print("\nNoun Chunks:")
for chunk in doc.noun_chunks:
    print(f'chunk.text : {chunk.text}')

# Iterate over named entities
print("\nNamed Entities:")
for ent in doc.ents:
    print(f"Entity: {ent.text}, Label: {ent.label_}")

print('Preprocessing done')
# Inspect tokens: our preprocessed document is now present as a list of tokens
# in our doc variable, and we can access its different annotations by looping through it

Vocabulary Object:

Sentences:
sent.text : La Joconde, ou Portrait de Mona Lisa, est un tableau de Léonard de Vinci, réalisé entre 1503 et 1506 ou entre 1513 et 1516[1],[2], et peut-être jusqu'à 1519 (l'artiste étant mort cette année-là, le 2 mai)[3], qui représente un portrait mi-corps, probablement celui de la Florentine Lisa Gherardini, épouse de Francesco del Giocondo.
sent.text : Acquise par François Ier, cette peinture à l'huile sur panneau de bois de peuplier de 77 × 53 cm est exposée au musée du Louvre à Paris. 


sent.text : Le Radeau de La Méduse est une peinture à l'huile sur toile, réalisée entre 1818 et 1819 par Théodore Géricault.


sent.text : La Nuit étoilée est une peinture réalisée par Vincent van Gogh.
sent.text : Le tableau représente ce que Van Gogh pouvait voir et extrapoler de la chambre qu'il occupait dans l'asile du monastère Saint-Paul-de-Mausole à Saint-Rémy-de-Provence en mai 1889.
sent.text : Souvent présenté comme son grand œuvre, le tableau a été reprodui

Do you see some issues with tokenisation?
 - Tokenization in spaCy can sometimes be tricky, especially when dealing with special characters, punctuation, or specific formatting. 

 - Above and below mechanisms helps to apply patterns to format with regular expression the expressivity.

#### Getting some help

Hint: You can either access Spacy's manual on the internet to find out how to access the information, or look at the built-in help by typing *help(doc)*.

https://spacy.io/api/doc

In [11]:
help(doc)

Help on Doc object:

class Doc(builtins.object)
 |  Doc(Vocab vocab, words=None, spaces=None, user_data=None, *, tags=None, pos=None, morphs=None, lemmas=None, heads=None, deps=None, sent_starts=None, ents=None)
 |  A sequence of Token objects. Access sentences and named entities, export
 |      annotations to numpy arrays, losslessly serialize to compressed binary
 |      strings. The `Doc` object holds an array of `TokenC` structs. The
 |      Python-level `Token` and `Span` objects are views of this array, i.e.
 |      they don't own the data themselves.
 |  
 |      EXAMPLE:
 |          Construction 1
 |          >>> doc = nlp(u'Some text')
 |  
 |          Construction 2
 |          >>> from spacy.tokens import Doc
 |          >>> doc = Doc(nlp.vocab, words=["hello", "world", "!"], spaces=[True, False, False])
 |  
 |      DOCS: https://spacy.io/api/doc
 |  
 |  Methods defined here:
 |  
 |  __bytes__(...)
 |      Doc.__bytes__(self)
 |  
 |  __getitem__(...)
 |      Get a `Token

#### Pandas

You can use Pandas to better visualize the results

In [12]:
# Using pandas for a better visualization
import pandas as pd

file_path = 'corpus_qa.txt'
with open( file_path  ) as infile:
  text = infile.read()
  
doc = nlp(text)
spacy_tokens = [w.text for w in doc]
data = pd.DataFrame(spacy_tokens, columns=['Word'])
data.head(7).style

,Word
0,La
1,Joconde
2,","
3,ou
4,Portrait
5,de
6,Mona


In [13]:
data["Word"].value_counts().sort_values(ascending=False)

Word
de          20
,           16
.           13
à           12
une         10
            ..
agit         1
équestre     1
Manuel       1
Godoy        1
\n           1
Name: count, Length: 221, dtype: int64

### 1.2 Sentence segmentation

**Exercise 5:**
Apart from token segmentation, Spacy has also automatically segmented our document into sentences.
* **(a)** Print out the different sentences of the document.
Hint: Look at the "Data descriptors " in the help page for 'doc'.



In [14]:
# Print the sentences
print("\nSentences:")
doc = nlp(text)

for sent in doc.sents:
    print(f'sent.text : {sent.text}')


Sentences:
sent.text : La Joconde, ou Portrait de Mona Lisa, est un tableau de Léonard de Vinci, réalisé entre 1503 et 1506 ou entre 1513 et 1516[1],[2], et peut-être jusqu'à 1519 (l'artiste étant mort cette année-là, le 2 mai)[3], qui représente un portrait mi-corps, probablement celui de la Florentine Lisa Gherardini, épouse de Francesco del Giocondo.
sent.text : Acquise par François Ier, cette peinture à l'huile sur panneau de bois de peuplier de 77 × 53 cm est exposée au musée du Louvre à Paris. 


sent.text : Le Radeau de La Méduse est une peinture à l'huile sur toile, réalisée entre 1818 et 1819 par Théodore Géricault.


sent.text : La Nuit étoilée est une peinture réalisée par Vincent van Gogh.
sent.text : Le tableau représente ce que Van Gogh pouvait voir et extrapoler de la chambre qu'il occupait dans l'asile du monastère Saint-Paul-de-Mausole à Saint-Rémy-de-Provence en mai 1889.
sent.text : Souvent présenté comme son grand œuvre, le tableau a été reproduit à de très nombreu

In [15]:
help(sent)

Help on Span object:

class Span(builtins.object)
 |  A slice from a Doc object.
 |  
 |  DOCS: https://spacy.io/api/span
 |  
 |  Methods defined here:
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getitem__(...)
 |      Get a `Token` or a `Span` object
 |      
 |      i (int or tuple): The index of the token within the span, or slice of
 |          the span to get.
 |      RETURNS (Token or Span): The token at `span[i]`.
 |      
 |      DOCS: https://spacy.io/api/span#getitem
 |  
 |  __gt__(self, value, /)
 |      Return self>value.
 |  
 |  __hash__(self, /)
 |      Return hash(self).
 |  
 |  __iter__(...)
 |      Iterate over `Token` objects.
 |      
 |      YIELDS (Token): A `Token` object.
 |      
 |      DOCS: https://spacy.io/api/span#iter
 |  
 |  __le__(self, value, /)
 |      Return self<=value.
 |  
 |  __len__(...)
 |      Get the number of tokens in the span.
 |      
 |      RETURN

## 3- POS tagging

Remember that the model corresponds to a processing 'pipeline' in Spacy:
  - by default, it includes the tokenisation, the lemmatization and the POS tagging

**Exercise**
- print each individual token, together with its lemmatized form and part of speech tag
- Use Panda to better visualize the results
- Look at the results, do you see any error?
- You can use the method 'spacy.explain' to have information about some annotation, for example the POS tags. Apply it to each POS tag to get a more detailed label.


In [49]:
# Define the columns you want to include in the DataFrame
columns = ["text", "lemma", "pos_", "explain Pos", "dependency", "dependency explained", "ent_type_", "Explain Ent_type", "is_stop"]
# Extract the token information
rows = []
for token in doc:
    row = [token.text, token.lemma_, token.pos_, spacy.explain(token.pos_), token.dep_ , spacy.explain(token.dep_), token.ent_type_, spacy.explain(token.ent_type_), token.is_stop]
    rows.append(row)

# Create the DataFrame
df = pd.DataFrame(rows, columns=columns)
# Print the DataFrame
df.head(10).style 

,text,lemma,pos_,explain Pos,dependency,dependency explained,ent_type_,Explain Ent_type,is_stop
0,Augusta,Augusta,PROPN,proper noun,compound,compound,PERSON,"People, including fictional",False
1,Ada,Ada,PROPN,proper noun,compound,compound,PERSON,"People, including fictional",False
2,King,King,PROPN,proper noun,nsubj,nominal subject,PERSON,"People, including fictional",False
3,",",",",PUNCT,punctuation,punct,punctuation,,None,False
4,Countess,Countess,PROPN,proper noun,appos,appositional modifier,PERSON,"People, including fictional",False
5,of,of,ADP,adposition,prep,prepositional modifier,,None,True
6,Lovelace,Lovelace,PROPN,proper noun,pobj,object of preposition,PERSON,"People, including fictional",False
7,(,(,PUNCT,punctuation,punct,punctuation,,None,False
8,née,née,PROPN,proper noun,compound,compound,,None,False
9,Byron,Byron,PROPN,proper noun,appos,appositional modifier,ORG,"Companies, agencies, institutions, etc.",False


#### Pandas

You can use Pandas to better visualize the results

In [17]:
# Using pandas for a better visualization
import pandas as pd

spacy_pos_tagged = [(w, w.lemma_, w.tag_, w.pos_) for w in doc]
pd.DataFrame(spacy_pos_tagged,
             columns=['Word', 'Lemma', 'POS tag', 'Tag type'])

,Word,Lemma,POS tag,Tag type
0,La,le,DET,DET
1,Joconde,Joconde,PROPN,PROPN
2,",",",",PUNCT,PUNCT
3,ou,ou,CCONJ,CCONJ
4,Portrait,portrer,VERB,VERB
...,...,...,...,...
402,de,de,ADP,ADP
403,Manuel,Manuel,PROPN,PROPN
404,Godoy,Godoy,PROPN,PROPN
405,.,.,PUNCT,PUNCT


#### Look at the results: do you see any issues?


 - Il y a des fautes de vocabulaire Portrait est interprété comme le verbe portrer du premier groupe à l'imparfait.

#### Notes on POS tags:

* You can use the method 'explain' to have information about some annotation, for example the POS tags, see the code below.
* Here we used a very small set of POS (vs e.g. 36 in the PTB: https://www.ling.upenn.edu/courses/Fall_2003/ling001/penn_treebank_pos.html)

In [18]:
# Inspect POS tags
all_tags = set()
for token in doc:
  all_tags.add(token.pos_)
for tag in all_tags:
  print( tag, spacy.explain(tag)) # explain each label

PROPN proper noun
DET determiner
ADV adverb
PUNCT punctuation
ADJ adjective
X other
NOUN noun
NUM numeral
VERB verb
CCONJ coordinating conjunction
ADP adposition
SPACE space
PRON pronoun
AUX auxiliary


## 5- Named entity recognition

As part of the preprocessing pipeline, Spacy has also carried out named entity recognition.

**Exercise 8:**
* print out each named entity, together with the label assigned to it
* what do the labels stand for?
* Use the module called 'displacy' to visualize the Named Entities directly in the text.

In [19]:
# Iterate over named entities
print("\nNamed Entities:")
for ent in doc.ents:
    print(f"Entity: {ent.text}, Label: {ent.label_} Explanation: {spacy.explain(ent.label_)}") 


Named Entities:
Entity: La Joconde, Label: PER Explanation: Named person or family.
Entity: Portrait de Mona Lisa, Label: MISC Explanation: Miscellaneous entities, e.g. events, nationalities, products or works of art
Entity: Léonard de Vinci, Label: PER Explanation: Named person or family.
Entity: Florentine Lisa Gherardini, Label: PER Explanation: Named person or family.
Entity: Francesco del Giocondo, Label: PER Explanation: Named person or family.
Entity: François Ier, Label: PER Explanation: Named person or family.
Entity: musée du Louvre, Label: LOC Explanation: Non-GPE locations, mountain ranges, bodies of water
Entity: Paris, Label: LOC Explanation: Non-GPE locations, mountain ranges, bodies of water
Entity: Le Radeau de La Méduse, Label: MISC Explanation: Miscellaneous entities, e.g. events, nationalities, products or works of art
Entity: Théodore Géricault, Label: PER Explanation: Named person or family.
Entity: La Nuit étoilée, Label: MISC Explanation: Miscellaneous entities

**Understand the Labels**
The labels assigned to named entities are part of a predefined set. Here are some common labels and their explanations:
 - PERSON: People, including fictional characters.
 - NORP: Nationalities or religious or political groups.
 - FAC: Buildings, airports, highways, bridges, etc.
 - ORG: Companies, agencies, institutions, etc.
 - GPE: Countries, cities, states.
 - LOC: Non-GPE locations, mountain ranges, bodies of water.
 - PRODUCT: Objects, vehicles, foods, etc. (not services).
 - MONEY: Monetary values, including unit.
 - DATE: Dates.
 - TIME: Times.
 - PERCENT: Percentages.
 - QUANTITY: Quantities, including counts and measurements.

In [20]:
from spacy import displacy
# Visualize the named entities using displaCy
displacy.render(doc, style="ent", jupyter=True)

Note on Named Entity Recognition: do you see any issues?


## 6- Parsing

Finally, as part of the pipeline, Spacy has also performed a dependency parsing (note that each module can de disabled if not needed).

**Exercise 9:** Extract all Noun Phrases from the file

* Retrieve the information from the dependency parses: dependent and head of each token
* Use displacy to visualize a parse tree: first try with a simple sentence (e.g. *La petite brise la glace.*) then use the first sentence of the document.
* Navigating the parse tree. Each element of the tree is associated to attributes: you can use them to inspect the different elements of the trees:
  * Define a Panda dataframe with each token id associated to its head, with the relation between them.
  * The eventual children of the current token are also printed.
* Print all the adjectives and the noun they modify



In [21]:
# Print for each token, its dependents and head
# Process the text to create a Doc object
text = "La petite brise la glace."
doc = nlp(text)

# Extract noun phrases
noun_phrases = list(doc.noun_chunks)
for phrase in noun_phrases:
    print(f'noun_chunks : {phrase.text}')

# Retrieve dependency parse information
for token in doc:
    print(f"Token: {token.text}, Dependency: {token.dep_}, Head: {token.head.text}, Head ID: {token.head.i}")

noun_chunks : La petite brise
noun_chunks : la glace
Token: La, Dependency: det, Head: brise, Head ID: 2
Token: petite, Dependency: amod, Head: brise, Head ID: 2
Token: brise, Dependency: ROOT, Head: brise, Head ID: 2
Token: la, Dependency: det, Head: glace, Head ID: 4
Token: glace, Dependency: nmod, Head: brise, Head ID: 2
Token: ., Dependency: punct, Head: brise, Head ID: 2


In [22]:
# Visualization using displacy
options = {
    "color": "blue",  # Color of the text (e.g., 'tomato')
    "font": "Source Sans Pro",  # Font of the text
    "bg": "white",  # Background color (e.g., light gray)
    "arrow_stroke": 1,  # Width of the arrows
    "arrow_width": 4,  # Width of the arrow head
    "word_spacing": 15,  # Space between words
    "distance": 80,  # Distance between words and arrows
    "fine_grained": True,  # Display fine-grained tags
    "compact": True,  # Compact display
    "font_size": "1.2em",  # Font size
    "image_width": "80px",  # Largeur de l'image
    "image_height": "10px"  # Hauteur de l'image
}
# Visualize the parse tree
displacy.render(doc, style="dep", options=options,jupyter=True)

#### Navigating the parse tree

In [23]:
# Navigating the parse tree
file_path = 'corpus_qa.txt'
with open( file_path  ) as infile:
  text = infile.read()

doc = nlp(text)
sentences = [ sent for sent in doc.sents]    

doc = nlp(sentences[0].text)
displacy.render(doc, style="dep", options=options, jupyter=True)

In [50]:
# Create a list to store the dependency information
dependency_info = []

# Iterate over the tokens and extract the dependency information
for token in doc:
    children = [child.text for child in token.children]
    dependency_info.append({
        "Token ID": token.i,
        "Token": token.text,
        "Head ID": token.head.i,
        "Head": token.head.text,
        "Dependency": token.dep_,
        "Explain" : spacy.explain(token.dep_),
        "Children": children
    })

# Create a Pandas DataFrame
df = pd.DataFrame(dependency_info)
df.head().style

,Token ID,Token,Head ID,Head,Dependency,Explain,Children
0,0,Augusta,2,King,compound,compound,[]
1,1,Ada,2,King,compound,compound,[]
2,2,King,26,was,nsubj,nominal subject,"['Augusta', 'Ada', ',', 'Countess', 'Byron', 'December', ',', 'known', ',']"
3,3,",",2,King,punct,punctuation,[]
4,4,Countess,2,King,appos,appositional modifier,['of']


 - ROOT: The root of the sentence.
 - nsubj: Nominal subject.
 - dobj: Direct object.
 - iobj: Indirect object.
 - csubj: Clausal subject.
 - ccomp: Clause complement.
 - xcomp: Open clausal complement.
 - pobj: Prepositional object.
 - dative: Dative object (used in some models).
 - prep: Prepositional modifier (used in some models).
 - advcl: Adverbial clause modifier.
 - acl: Clause modifier.
 - amod: Adjectival modifier.
 - det: Determiner.
 - aux: Auxiliary.
 -  cop: Copula.
 -  conj: Conjunct.

La Joconde, ou Portrait de Mona Lisa, est un tableau de Léonard de Vinci, réalisé entre 1503 et 1506 ou entre 1513 et 1516[1],[2], et peut-être jusqu'à 1519 (l'artiste étant mort cette année-là, le 2 mai)[3], qui représente un portrait mi-corps, probablement celui de la Florentine Lisa Gherardini, épouse de Francesco del Giocondo.


In [25]:

# Print all adjectives and the nouns they modify
for token in doc:
    token_head = token.head 
    print(f'token_head  : {token_head.text}')
    if token_head.pos_ == "ADJ" or token_head.pos_ == "NOUN" :
       for child in token_head.children:
         #  if (child.dep_ == "amod" or child.dep_ == "nmod") and (child.pos_ == "ADJ" or child.pos_ == "NOUN") :
           if (child.dep_ == "amod" and child.pos_ == "ADJ")  or  (child.dep_ == "nmod" and child.pos_ == "NOUN") :
               print(f"Find {token_head.text} impacts {child.text} as {child.dep_}")

token_head  : Joconde
token_head  : tableau
Find tableau impacts artiste as nmod
Find tableau impacts mai)[3 as nmod
token_head  : Joconde
token_head  : Portrait
token_head  : Joconde
token_head  : Mona
token_head  : Portrait
token_head  : Mona
token_head  : tableau
Find tableau impacts artiste as nmod
Find tableau impacts mai)[3 as nmod
token_head  : tableau
Find tableau impacts artiste as nmod
Find tableau impacts mai)[3 as nmod
token_head  : tableau
Find tableau impacts artiste as nmod
Find tableau impacts mai)[3 as nmod
token_head  : tableau
Find tableau impacts artiste as nmod
Find tableau impacts mai)[3 as nmod
token_head  : Léonard
token_head  : tableau
Find tableau impacts artiste as nmod
Find tableau impacts mai)[3 as nmod
token_head  : Vinci
token_head  : Léonard
token_head  : tableau
Find tableau impacts artiste as nmod
Find tableau impacts mai)[3 as nmod
token_head  : tableau
Find tableau impacts artiste as nmod
Find tableau impacts mai)[3 as nmod
token_head  : 1503
token_h

In [26]:
# Itérer sur les tokens et trouver les amod et nmod
for token in doc:
    if token.dep_ == "amod":
        # Remonter sur le mot impacté
        head_token = token.head
        print(f"Token : {token.text}, POS Tag : {token.pos_}, Dependency : {token.dep_} Mot impacté : {head_token.text}, POS Tag : {head_token.pos_}, Dependency : {head_token.dep_}")
    elif token.dep_ == "nmod":
        # Remonter sur le mot impacté
        head_token = token.head
        print(f"Token : {token.text}, POS Tag : {token.pos_}, Dependency : {token.dep_} Mot impacté : {head_token.text}, POS Tag : {head_token.pos_}, Dependency : {head_token.dep_}")

Token : Mona, POS Tag : PROPN, Dependency : nmod Mot impacté : Portrait, POS Tag : VERB, Dependency : conj
Token : Léonard, POS Tag : PROPN, Dependency : nmod Mot impacté : tableau, POS Tag : NOUN, Dependency : ROOT
Token : Vinci, POS Tag : NOUN, Dependency : nmod Mot impacté : Léonard, POS Tag : PROPN, Dependency : nmod
Token : artiste, POS Tag : NOUN, Dependency : nmod Mot impacté : tableau, POS Tag : NOUN, Dependency : ROOT
Token : mai)[3, POS Tag : NOUN, Dependency : nmod Mot impacté : tableau, POS Tag : NOUN, Dependency : ROOT
Token : mi-corps, POS Tag : NOUN, Dependency : amod Mot impacté : portrait, POS Tag : NOUN, Dependency : obj
Token : Francesco, POS Tag : NOUN, Dependency : nmod Mot impacté : épouse, POS Tag : NOUN, Dependency : appos
Token : Giocondo, POS Tag : PROPN, Dependency : nmod Mot impacté : Francesco, POS Tag : NOUN, Dependency : nmod


## 7- Putting it all together

Now we are going to use the skills practiced in the preceding exercises to build a simple question-answering system on a toy dataset (in French).

We will focus on specific questions of the form "Qui a peint X ?".
We will define patterns based on differents ways of formulating this question, and use them to extract the answer from a small toy corpus based on wikipedia pages on paintings.

When you're done with this exercise, try to answer other types of questions, such as "Où est exposé X ?", "Quand a été peinte X ?".

Below, we reload the spacy French model adding specific options to merge named entities containing multiple tokens.

**Exercise 10:** : In this part, we focus on the first question. 
- This question is designed to be similar to the document containing the answer. 
- We can thus define a pattern based on its structure to extract the answer from the document.

- Process the question using the spacy nlp pipeline
- display its parse tree and / or print a Pandas dataframe containing information from the parse tree
- Now to define a lexico-syntactic pattern, you are going to use spacy *DependencyMatcher* : https://spacy.io/usage/rule-based-matching#dependencymatcher
  * Look at the doc to understand how it works
  * Define a pattern that should match 'qui ' in the question

::: 

Here is the list of questions we will consider.
You also need a corpus of source document, you can find it on Moodle (corpus_qa.txt).

:::

:::

 - **Qui, précédé d'une préposition,(de ,pour, par (complément d'agent)) est complément d'objet indirect, complément circonstanciel, complément d'agent, complément du nom ou de l'adjectif, et son antécédent ne peut être qu'une personne (lorsque l'antécédent est une chose, on emploie les pronoms relatifs Lequel, Laquelle, Lesquels, Lesquelles ou les formes)** ...

 :::

 :::

 - **Le complément d'agent est un complément qui n'apparaît qu'à la voix passive.** 
 - **Il est toujours introduit par la préposition par ou de.**   
 - **Le complément d'agent indique la personne ou la chose qui effectue l'action du verbe**. 

 :::

In [27]:
# Load the model
# Navigating the parse tree
question_list = [
    'La Joconde est un tableau de qui ?',
    'Le radeau de la méduse est une peinture réalisée par qui ?'
]
file_path = 'corpus_qa.txt'
with open( file_path  ) as infile:
  text = infile.read()

nlp = spacy.load('fr_core_news_sm')
nlp.add_pipe("merge_entities")
nlp.add_pipe("merge_noun_chunks")

doc = nlp(text)

doc_requst0 = nlp(question_list[0])
doc_requst1 = nlp(question_list[1])

In [28]:

displacy.render(doc_requst0, style="dep", jupyter=True,options=options)
displacy.render(doc_requst1, style="dep", jupyter=True,options=options)  
#for sent in doc.sents:
#    displacy.render(sent, style="dep", jupyter=True,options=options) 

In [29]:
import re

def get_nounchunk(doc,target_token):
    for chunk in doc.noun_chunks:
        if target_token.lower() in chunk.text.lower():
            return chunk.text
    return target_token        

def forme2_qui (doc):
    pattern = r'(de|pour|par)\s+qui\s*[?!]'
    match = re.search(pattern, doc.text)
    if match:
       return True
    return False   

def retrieve_nmod_nsubj(docWhole, docreq):
    for sentence in docreq.sents:
        root_token = sentence.root
        for child in root_token.children:
            if child.dep_ == "nsubj" : 
               value = get_nounchunk(docWhole, child.text) 
               if value != "NONE" :
                  return root_token.text, value
               return root_token.text, child.text                                             

def who_realized(doc, quoi, dequi ):
    for token in doc:
        if token.head.text == quoi  and dequi in token.text :
            token_head = token.head
            for child in token_head.children:
                if  child.dep_ == "nmod" and child.pos_ == "PROPN": # Trouver le sujet du verbe
                    return  child.text
    return "NONE"  

doc = nlp(text)

for question in question_list:
    doc_requst = nlp(question)
    if forme2_qui (doc_requst):
       quoi, dequi = retrieve_nmod_nsubj(doc, doc_requst)
       print(f' quoi : {quoi} de qui : {dequi} par qui : {who_realized(doc, quoi, dequi) }')  


 quoi : tableau de qui : La Joconde par qui : Léonard de Vinci
 quoi : peinture de qui : Le Radeau de La Méduse par qui : NONE


In [30]:
# Display the parse tree of the first question
# Navigating the parse tree
file_path = 'corpus_qa.txt'

with open( file_path  ) as infile:
  text_corpus = infile.read()

doc_corpus = nlp(text_corpus)
doc_requst = nlp(question_list[0])

sentences_corpus  = [ sent for sent in doc_corpus.sents]  
sentences_request = [ sent for sent in doc_requst.sents]    

docorp = nlp(doc_corpus.text)
docreq = nlp(question_list[0])
displacy.render(docorp, style="dep", options=options, jupyter=True)
displacy.render(docreq, style="dep", options=options, jupyter=True)

**Exercise**:
Retrieve the documents that are relevant to the question, i.e. the ones containing the keyword 'La Joconde'.

It is recommended to define a function, that could be used for the next exercises. It could work by:

* first indexing all the documents using the named entities present in the document (i.e. build a dictionnary mapping a named entity to all documents where it is present)
* now the *retrieve_documents(...)* method should try to match the input keywork with a named entity and return the matching documents.
* Test the function with 'Joconde':
  * Does it work with the method based on named entities?
  * Add a backup solution : if no document is found,simply try to find the string corresponding to the keyword in the document

 - **Mots interrogatifs** : L'utilisation de mots interrogatifs spécifiques comme qui, que, quoi, comment, pourquoi, où, quand, etc.
  - **Position des mots interrogatifs** : Ces mots peuvent se placer en début ou en fin de phrase selon le style et l'emphase désirée.

In [31]:
import spacy
import re

# Charger le modèle français
nlp = spacy.load("fr_core_news_sm")

# Liste des mots interrogatifs
interrogative_words = {'comment', 'laquelle', 'lequel', 'lesquelles', 'lesquels', 'où', 'pourquoi', 'quand', 'que', 'qui', 'quoi', 'quel', 'quelle', 'quels', 'quelles'}

def analyser_phrase_interrogative_(phrase,docWhole):
    doc = nlp(phrase)
    elements = {
        'mot_interrogatif': None,
        'verbe': None,
        'racine' : None,
        'sujet': None,
        'complement': None,
        'chunk' : None,
        'point_interrogation': doc[-1].text == "?"
    }

    def get_nounchunk(doc,target_token):
        for chunk in doc.noun_chunks:
            if target_token.lower() in chunk.text.lower():
                return chunk.text
        return target_token  

    for sentence in doc.sents:
        root_token = sentence.root
    #    print(f'root_token {root_token}')
        elements['racine'] = str(root_token)
        for token in doc:
    #        print(f'token {token} {token.head}  {token.dep_}  {token.lemma_}')
            if token.pos_ == "PRON" and token.lemma_.lower() in {'qui', 'que', 'quoi', 'quel', 'quelle', 'quels', 'quelles', 'où', 'comment', 'pourquoi', 'quand'}:
                elements['mot_interrogatif'] = token.lemma_.lower()
            elif token.pos_ == "VERB" and not elements['verbe']:
                elements['verbe'] = token.lemma_
            elif token.dep_ in {'nsubj','nsubj:pass'} and not elements['sujet']:
                elements['sujet'] = token.text 
            elif token.dep_ in ["nmod"] and not elements['complement']:
                elements['complement'] = token.text
                elements['chunk'] = get_nounchunk(docWhole, token.head.text)  
        return elements

doc = nlp(text)    
#question = "La Joconde a été peinte par qui ?"
question="Le Radeau de la Méduse est une peinture réalisée par qui ?"
elements = analyser_phrase_interrogative_(question,doc)  
elements  

{'mot_interrogatif': 'qui',
 'verbe': 'réaliser',
 'racine': 'peinture',
 'sujet': 'Radeau',
 'complement': 'Méduse',
 'chunk': 'Le Radeau',
 'point_interrogation': True}

In [32]:
displacy.render(nlp(question), style="dep", options=options,jupyter=True)

__analyser_elements__ :   
 - Cette fonction extrait les mots-clés importants de la question analysée.   
 - Elle parcourt les éléments, excluant le point d'interrogation, et ajoute les valeurs non nulles à la liste des mots-clés.

In [33]:
def analyser_elements(elements):
    mots_cles = []
    #print(f'elements: {elements}')
    for key, value in elements.items():
        if value and key not in ['mot_interrogatif', 'point_interrogation']:
            mots_cles.append((key,value))
    return mots_cles

mots_cles = analyser_elements(elements)
mots_cles

[('verbe', 'réaliser'),
 ('racine', 'peinture'),
 ('sujet', 'Radeau'),
 ('complement', 'Méduse'),
 ('chunk', 'Le Radeau')]

In [34]:
import string

def nettoyer_phrase(phrase):
    # Créer une liste de caractères en excluant la ponctuation, puis la rejoindre
    return ''.join(char for char in phrase if char not in string.punctuation)

def comparer_elements(phrases, elements):   
    correspondances = []
    elements = dict(elements)
    for phrase_ in phrases:
        phrase = nettoyer_phrase(str(phrase_))
        score = 0  # Initialiser le score pour chaque phrase
        words_in_phrase = phrase.lower().split()  # Convertir la phrase en mots et en minuscules
       # print(words_in_phrase)
        
        # Vérifier la correspondance du verbe
        if 'verbe' in elements and elements['verbe'] in words_in_phrase:
            score += 1  # Ajouter 1 au score pour le verbe
        
        # Vérifier la correspondance du sujet
        if 'racine' in elements and elements['racine'].lower() in words_in_phrase:
            score += 1  # Ajouter 1 au score pour le sujet
        
        # Vérifier la correspondance de l'objet
        if 'sujet' in elements and elements['sujet'].lower() in words_in_phrase:
            score += 1  # Ajouter 1 au score pour l'objet
        
        # Vérifier la correspondance des compléments
        if 'complement' in elements :
            for complement in elements['complement']:    
                if complement.lower() in words_in_phrase:
                    score += 1  # Ajouter 1 au score pour chaque complément trouvé
        
        correspondances.append((phrase, score))  # Ajouter la phrase et son score
    return correspondances

def filtrer_phrases_maximum(correspondances):
    # Trouver le score maximum
    max_score = max(score for _, score in correspondances)
    # Filtrer les phrases avec le score maximum
    phrases_max = [phrase for phrase, score in correspondances if score == max_score]
    return phrases_max

doc_corpus = nlp(text_corpus)
sentences_corpus  = [ sent for sent in doc_corpus.sents] 
texte_source = filtrer_phrases_maximum(comparer_elements(sentences_corpus, mots_cles))
print(f'texte_source : {texte_source}')
print(f'mots_cles : {mots_cles}')

texte_source : ['Le Radeau de La Méduse est une peinture à lhuile sur toile réalisée entre 1818 et 1819 par Théodore Géricault\n\n']
mots_cles : [('verbe', 'réaliser'), ('racine', 'peinture'), ('sujet', 'Radeau'), ('complement', 'Méduse'), ('chunk', 'Le Radeau')]


In [35]:
import spacy
from spacy.matcher import DependencyMatcher

nlp = spacy.load("fr_core_news_sm")

def resolve_request(question, doc, pattern):
    elements = analyser_phrase_interrogative_(question,doc)  
    mots_cles = analyser_elements(elements)
    sentences_corpus  = [ sent for sent in doc_corpus.sents] 
    texte_source = filtrer_phrases_maximum(comparer_elements(sentences_corpus, mots_cles))
    #print(f'-- {question} -- {texte_source}')
    matcher = DependencyMatcher(nlp.vocab)
    matcher.add("test_pattern", pattern)

    doc = nlp(str(texte_source))
    matches = matcher(doc)

    for match_id, token_ids in matches:
        pattern_name = nlp.vocab.strings[match_id]
        matched_tokens = [doc[i] for i in token_ids]
        print(f"Pattern '{pattern_name}' matched: {matched_tokens}")  
    return texte_source         

**Exercise 12:** Test the pattern
- Apply the pattern to each sentence of the retrieved document, do you find the right answer?

In [36]:
leonard_pattern = [
        {
            "RIGHT_ID": "start",
            "RIGHT_ATTRS": {"POS": {"IN": ["ADP","PROPN"]}}
        },
        {
            "LEFT_ID": "start",
            "REL_OP": ">",
            "RIGHT_ID": "agent",
            "RIGHT_ATTRS": {"DEP": {"IN": ["nmod","advcl"]}}
        }
    ]

questionA = "La Joconde a été peinte par qui ?"
texte_source = resolve_request(questionA, nlp(text_corpus), [leonard_pattern])

Pattern 'test_pattern' matched: [Léonard, Vinci]


In [37]:
# Display the parse tree of the second question
displacy.render(nlp(str(questionA)), style="dep", options=options,jupyter=True)
displacy.render(nlp(str(texte_source)), style="dep", options=options,jupyter=True)

**Exercise 13:** Now define a pattern to answer the second question, and find the answer!

Be careful, there is a little issue here:
- Display the parse tree for the question and the matching document: what do you observe?
- Build a pattern to match the question and the answer (Hint : you can use a list of dependency relations in the pattern, using e.g. *"RIGHT_ATTRS": {"DEP": {"IN":[ "acl", "advcl" ] }*
- Finally, retrieve the answer

In [38]:
radeau_pattern = [
        {
            "RIGHT_ID": "start",
            "RIGHT_ATTRS": {"POS": {"IN": ["ADP","PROPN"]}}
        },
        {
            "LEFT_ID": "start",
            "REL_OP": ">++",
            "RIGHT_ID": "agent",
            "RIGHT_ATTRS": {"DEP": {"IN": ["flat:name"]}}
        },
        {
            "LEFT_ID": "agent",
            "REL_OP": "<",
            "RIGHT_ID": "drole",
            "RIGHT_ATTRS": {"DEP": {"IN": ["obl:mod"]}}
        }

    ]

questionB ="Le Radeau de la Méduse est une peinture réalisée par qui ?"
texte_source = resolve_request(questionB, nlp(text_corpus), [radeau_pattern])

Pattern 'test_pattern' matched: [Théodore, Géricault\n\n, Théodore]


In [39]:
# Display the parse tree of the second question
displacy.render(nlp(str(question)), style="dep", options=options,jupyter=True)
displacy.render(nlp(str(texte_source)), style="dep", options=options,jupyter=True)

In [40]:
#  Apply the pattern to each sentence of this document, do you find the right answer?
#  Define a pattern that matches both the question and answer
#  Here we have a slight difference between the dep rel label for quest and answer

request=[questionA,questionB]
pattern_peinture = [leonard_pattern, radeau_pattern]

for question in request :
    resolve_request(question, nlp(text_corpus), pattern_peinture)

Pattern 'test_pattern' matched: [Léonard, Vinci]
Pattern 'test_pattern' matched: [Théodore, Géricault\n\n, Théodore]


**Exercise** Use the patterns defined to find the answers to the questions below. Here, we know that we're looking for the painter, we don'y want to match the question to the answer, we want to test known patterns to find the right answer.

- find a way to extract the name of the painting from the question
- retrieve the relevant document
- test the patterns defined previously to extract the correct answer

In [41]:
new_questions = [
    'Qui a peint American Gothic ?',
    'Qui est l\'auteur de la peinture La Nuit étoilée ?',
    'Qui a réalisé Un Garrochista ?',
]

In [42]:
def make_entity(new_questions):
    entity=[]
    for question in new_questions:
        doc = nlp(question)
        for ent in doc.ents:
    #        print(f"Question: {question} Entity: {ent.text}, Label: {ent.label_}")
            entity.append((nettoyer_phrase(str(ent.text)),ent.label_))
    return entity

In [43]:
import spacy
from spacy.pipeline import EntityRuler

# Charger le modèle français
nlp = spacy.load("fr_core_news_sm")

# Créer un EntityRuler
ruler = nlp.add_pipe("entity_ruler", before="ner")

# Définir les patterns d'entités
patterns = [
    {"label": "ARTWORK", "pattern": "American Gothic"},
    {"label": "ARTWORK", "pattern": "La Nuit étoilée"},
    {"label": "ARTWORK", "pattern": "Un Garrochista"}
]
# Ajouter les patterns à l'EntityRuler
ruler.add_patterns(patterns)

In [44]:
# Recherche d'entités dans le corpus
def rechercher_phrases(entity_name,sentences_corpus, label):
    for document in sentences_corpus:
        doc = nlp(str(document))
        for ent in doc.ents:
            if ent.label_ == label:
                print(f"Document: {document}\nEntity found: {ent.text} (Type: {ent.label_})\n")

In [45]:
doc_corpus = nlp(text_corpus)
sentences_corpus  = [ sent for sent in doc_corpus.sents] 
entity = make_entity(new_questions)
rechercher_phrases(entity,sentences_corpus,"ARTWORK")

Document: La Nuit étoilée est une peinture réalisée par Vincent van Gogh.
Entity found: La Nuit étoilée (Type: ARTWORK)

Document: Il est maintenant conservé dans le Museum of Modern Art (MOMA) à New York depuis 1941[1].

American Gothic est un tableau de Grant Wood appartenant à la Friends of American
Entity found: American Gothic (Type: ARTWORK)

Document: Un Garrochista est une peinture réalisée par Francisco de Goya en 1795.
Entity found: Un Garrochista (Type: ARTWORK)



**Exercise** To go further

Try to:
- find who painted 'Le Cri' and 'Arearea
- where are located 'La Joconde' and 'Arearea'

In [46]:
new_questions_A = [
    "Qui a peint 'Le Cri' et Areaarea ?",
    "Ou est localisé 'La Joconde' et Areaarea"
]
# Définir les patterns d'entités
patterns_A = [
    {"label": "ART_WORK", "pattern": "Le Cri"},
    {"label": "ART_WORK", "pattern": "La Joconde"}
]
# Ajouter les patterns à l'EntityRuler
ruler.add_patterns(patterns_A)

doc_corpus = nlp(text_corpus)
sentences_corpus  = [ sent for sent in doc_corpus.sents] 
entity = make_entity(new_questions_A)
rechercher_phrases(entity,sentences_corpus,"ART_WORK")

Document: La Joconde, ou Portrait de Mona Lisa, est un tableau de Léonard de Vinci, réalisé entre 1503 et 1506 ou entre 1513 et 1516[1],[2], et peut-être jusqu'à 1519 (l'artiste étant mort cette année-là, le 2 mai)[3], qui représente un portrait mi-corps, probablement celui de la Florentine Lisa Gherardini, épouse de Francesco del Giocondo.
Entity found: La Joconde (Type: ART_WORK)

Document: Le Cri (en norvégien : Skrik) est une œuvre expressionniste de l'artiste norvégien Edvard
Entity found: Le Cri (Type: ART_WORK)

